src/SignalR/clients/ts/signalr/src/HubConnection.ts

```diff
@@ -101,48 +118,110 @@ export class HubConnection {
      * @returns {Promise<void>} A Promise that resolves when the connection has been successfully established, or rejects with an error.
      */
     public async start(): Promise<void> {
-        const handshakeRequest: HandshakeRequestMessage = {
-            protocol: this.protocol.name,
-            version: this.protocol.version,
-        };
+        if (this.connectionState !== HubConnectionState.Disconnected) {
+            return Promise.reject(new Error("Cannot start a HubConnection that is not in the 'Disconnected' state."));
+        }
 
+        this.connectionState = HubConnectionState.Connecting;
         this.logger.log(LogLevel.Debug, "Starting HubConnection.");
 
+        try {
+            await this.startInternal();
+
+            this.connectionState = HubConnectionState.Connected;
+            this.logger.log(LogLevel.Debug, "HubConnection connected successfully.");
+        } catch (e) {
+            this.connectionState = HubConnectionState.Disconnected;
+            this.logger.log(LogLevel.Debug, `HubConnection failed to start successfully because of error '${e}'.`);
+            return Promise.reject(e);
```

Comment:
Because this is in an async method, should this be a throw instead of `Promise.reject(e);`

Reply:
I was wondering this myself. Our TypeScript client switches back-and-forth between throwing and returning rejected promises from async methods.

The only difference I can ascertain between the two is that throwing will trigger the catch of a surrounding try/catch *in the same method*. Does anyone else know of any other differences?

I switch a few returned rejected promises to throws in HttpConnection.startInternal where I need the catch block to run. Everywhere else I try to return a rejected promise since that seems to be the more common patter in our client.

Reply:
Something to figure out, but not in this PR 😬